src/hardware/innovation.cpp

```diff
@@ -0,0 +1,315 @@
+	                                   sid_sample_rate, passband);
+
+	// Setup and assign the port address
+	const auto read_from = std::bind(&Innovation::ReadFromPort, this, _1, _2);
+	const auto write_to = std::bind(&Innovation::WriteToPort, this, _1, _2, _3);
+	base_port = static_cast<uint16_t>(port_choice);
+	read_handler.Install(base_port, read_from, IO_MB, 0x20);
+	write_handler.Install(base_port, write_to, IO_MB, 0x20);
+
+	// Move the locals into members
+	service = std::move(sid_service);
+	channel = std::move(mixer_channel);
+
+	// Ready state-values for rendering
+	last_used = 0;
+	play_buffer_pos = 0;
+	keep_rendering = true;
+
+	// Start rendering
+	renderer = std::thread(std::bind(&Innovation::Render, this));
+	set_thread_name(renderer, "dosbox:innovatn"); // < 16-character cap
+	play_buffer = playable.Dequeue(); // populate the first play buffer
+
+	if (filter_strength == 0)
+		LOG_MSG("INNOVATION: Running on port %xh with a SID %s at %0.3f MHz",
+		        base_port, model_choice.c_str(), chip_clock / 1000000.0);
+	else
+		LOG_MSG("INNOVATION: Running on port %xh with a SID %s at %0.3f MHz filtering at %d%%",
+		        base_port, model_choice.c_str(), chip_clock / 1000000.0,
+		        filter_strength);
+
+	is_open = true;
+}
+
+void Innovation::Close()
+{
+	if (!is_open)
+		return;
+
+	DEBUG_LOG_MSG("INNOVATION: Shutting down the SSI-2001 on port %xh", base_port);
+
+	// Stop playback
+	if (channel)
+		channel->Enable(false);
+
+	// Stop rendering and drain the queues
+	keep_rendering = false;
+	if (!backstock.Size())
+		backstock.Enqueue(std::move(play_buffer));
+	while (playable.Size())
+		play_buffer = playable.Dequeue();
+
+	// Wait for the rendering thread to finish
+	if (renderer.joinable())
+		renderer.join();
+
+	// Remove the IO handlers before removing the SID device
+	read_handler.Uninstall();
+	write_handler.Uninstall();
+
+	// Reset the members
+	channel.reset();
+	service.reset();
+	is_open = false;
+}
+
+size_t Innovation::ReadFromPort(size_t port, MAYBE_UNUSED size_t iolen)
+{
+	const auto sid_port = static_cast<uint16_t>(port - base_port);
+	const std::lock_guard<std::mutex> lock(service_mutex);
+	return service->read(sid_port);
+}
+
+void Innovation::WriteToPort(size_t port, size_t val, MAYBE_UNUSED size_t iolen)
+{
+	assert(val <= UINT8_MAX);
+	const auto data = static_cast<uint8_t>(val); // 8-bit IO
+	const auto sid_port = static_cast<uint16_t>(port - base_port);
+	{ // service-lock
+		const std::lock_guard<std::mutex> lock(service_mutex);
+		service->write(sid_port, data);
+	}
+	// Turn on the channel after the data's written
+	if (!last_used) {
+		channel->Enable(true);
+	}
+	last_used = PIC_Ticks;
+}
+
+void Innovation::Render()
+{
+	const auto cycles_per_sample = static_cast<uint16_t>(chip_clock /
+	                                                     sid_sample_rate);
+
+	// Allocate one buffer and reuse it for the duration.
+	std::vector<int16_t> buffer(SAMPLES_PER_BUFFER);
```

Comment:
Since `SAMPLES_PER_BUFFER` is known at compile-time, is there any efficiency (memory or performance) gain from `std::array<int16_t, SAMPLES_PER_BUFFER> buffer;` instead?